src/diffusers/models/unet_2d_blocks.py

```diff
@@ -2194,11 +2212,34 @@ def forward(
         encoder_attention_mask: Optional[torch.FloatTensor] = None,
     ):
         lora_scale = cross_attention_kwargs.get("scale", 1.0) if cross_attention_kwargs is not None else 1.0
+        is_freeu_enabled = (
+            getattr(self, "s1", None)
+            and getattr(self, "s2", None)
+            and getattr(self, "b1", None)
+            and getattr(self, "b2", None)
+        )
 
         for resnet, attn in zip(self.resnets, self.attentions):
             # pop res hidden states
             res_hidden_states = res_hidden_states_tuple[-1]
             res_hidden_states_tuple = res_hidden_states_tuple[:-1]
+
+            # Courtesy:
+            # https://github.com/ChenyangSi/FreeU
+            # https://github.com/lyn-rgb/FreeU_Diffusers
+            if is_freeu_enabled is not None:
+                # --------------- FreeU code -----------------------
+                # Only operate on the first two stages
+                if self.resolution_idx == 0:
+                    num_half_channels = hidden_states.shape[1] // 2
+                    hidden_states[:, :num_half_channels] = hidden_states[:, :num_half_channels] * self.b1
+                    res_hidden_states = fourier_filter(res_hidden_states, threshold=1, scale=self.s1)
+                if self.resolution_idx == 1:
+                    num_half_channels = hidden_states.shape[1] // 2
+                    hidden_states[:, :num_half_channels] = hidden_states[:, :num_half_channels] * self.b2
+                    res_hidden_states = fourier_filter(res_hidden_states, threshold=1, scale=self.s2)
+                # ---------------------------------------------------------
```

Comment:
```suggestion
                 hidden_states, residual_hidden_states = self.apply_freeu(hidden_states, res_hidden_states, self.b1, self.b2, self.s1, self.s2R)
```

Could we move this and all the other code into a `apply_freeu` function to keep the already bloated unet code a bit smaller?